### PR TITLE
storage: fetch blob.meta in asynchronous mode

### PR DESCRIPTION
**Pull request by jiangliu**
Fetch blob.meta in asynchronous mode, to reduce container startup time.

Signed-off-by: Jiang Liu <gerry@linux.alibaba.com>

### CODE REVIEW

**Review by imeoer — APPROVED** [review]

**Comment by changweige** [comment]
Unfortunately, I observed a performance regression with the PR applied to main branch :-(

**Comment by jiangliu** [comment]
> Unfortunately, I observed a performance regression with the PR applied to main branch :-(

That doesn't make sense:)
It should not cause so big regression, could you please help to retest?

I could always observe performance improvements in my env.

File: storage/src/cache/cachedfile.rs [hunk 1]
```diff
@@ -36,11 +37,69 @@ use crate::meta::{BlobMetaChunk, BlobMetaInfo};
 use crate::utils::{alloc_buf, copyv, readv, MemSliceCursor};
 use crate::{StorageError, StorageResult, RAFS_DEFAULT_CHUNK_SIZE};
 
+const DOWNLOAD_META_RETRY_COUNT: u32 = 5;
+const DOWNLOAD_META_RETRY_DELAY: u64 = 500;
+
+#[derive(Default, Clone)]
+pub(crate) struct FileCacheMeta {
+    status: Arc<AtomicBool>,
```
Comment (liubogithub) [inline storage/src/cache/cachedfile.rs:45]: Do you mind adding a comment for `status`? 
looks like `status` indicates the failure of downloading blob.meta

**Comment by changweige** [comment]
> > Unfortunately, I observed a performance regression with the PR applied to main branch :-(
> 
> That doesn't make sense:) It should not cause so big regression, could you please help to retest?
> 
> I could always observe performance improvements in my env.

I must be too sleepy that night. 😹 Yes performance with this PR. Sorry for the noise I made

File: storage/src/cache/cachedfile.rs [hunk 2]
```diff
@@ -137,17 +196,26 @@ impl BlobCache for FileCacheEntry {
...
-                    let cki = BlobMetaChunk::new(chunk_index as usize, &chunks_meta.state);
-                    b.chunkinfo = BlobIoChunk::Base(Arc::new(cki));
+            if let BlobIoChunk::Address(_blob_index, chunk_index) = b.chunkinfo {
+                if let Some(meta) = self.meta.as_ref() {
+                    if let Some(bm) = meta.get_blob_meta() {
+                        let cki = BlobMetaChunk::new(chunk_index as usize, &bm.state);
+                        b.chunkinfo = BlobIoChunk::Base(Arc::new(cki));
+                    } else {
+                        warn!("failed to get blob.meta for prefetch");
+                        fail = true;
```
Comment (changweige) [inline storage/src/cache/cachedfile.rs:208]: Can we just return error from here>

File: storage/src/cache/cachedfile.rs [hunk 3]
```diff
@@ -36,11 +37,69 @@ use crate::meta::{BlobMetaChunk, BlobMetaInfo};
 use crate::utils::{alloc_buf, copyv, readv, MemSliceCursor};
 use crate::{StorageError, StorageResult, RAFS_DEFAULT_CHUNK_SIZE};
 
+const DOWNLOAD_META_RETRY_COUNT: u32 = 5;
+const DOWNLOAD_META_RETRY_DELAY: u64 = 500;
```
Comment (changweige) [inline storage/src/cache/cachedfile.rs:41]: totally waiting for 2.5 seconds in a terrible network environment could not be enough. Can we make it longer like 5 seconds or 10 seconds?

File: storage/src/cache/cachedfile.rs (hunk 1 above)
Comment (jiangliu) [reply storage/src/cache/cachedfile.rs:45]: renamed to `has_error`.

File: storage/src/cache/cachedfile.rs (hunk 2 above)
Comment (jiangliu) [reply storage/src/cache/cachedfile.rs:208]: We are within an "for_each()" loop and can't return error here.

File: storage/src/cache/cachedfile.rs (hunk 3 above)
Comment (jiangliu) [reply storage/src/cache/cachedfile.rs:41]: changed to 20s.

Comment (changweige) [reply storage/src/cache/cachedfile.rs:41]: Looks it was not changed, forget to stage changes?

Comment (jiangliu) [reply storage/src/cache/cachedfile.rs:41]: Sorry, changed to 10s instead of 20s:)

**Pull request merged**